torch/utils/data/datapipes/datapipe.py

```diff
@@ -154,6 +156,12 @@ def __getstate__(self):
             return IterDataPipe.getstate_hook(self)
         return self.__dict__
 
+    def __setstate__(self, state):
+        """
+        This method definition is necessary so that the hook function in `_typing.py` can work.
+        """
+        self.__dict__ = state
```

Comment:
I think not all `super()` has `__setstate__`. I can change it to your first suggestion.

Reply:
You can do a quick experiment. I believe Python class always has a `__setstate__` function hidden from the scene

Reply:
It doesn't seem to be the case. I changed the function to:
```python
def __setstate__(self, state):
    """
    This method definition is necessary so that the hook function in `_typing.py` can work.
    """
    return super().__setstate__(state)
```
Then, I ran this code snippet:
```python
dp = IterableWrapper(range(10))
serialized_dp = pickle.dumps(dp)
deserialized_dp = pickle.loads(serialized_dp)
print(list(dp))
```
It raises:
```
AttributeError: 'super' object has no attribute '__setstate__'
```

Reply:
I think `self.__dict__.update(state)` is likely the best option (aside from `self.__dict__ = state`)